application/common/models/User.php

```diff
@@ -0,0 +1,190 @@
+<?php
+
+namespace common\models;
+
+use common\helpers\Utils;
+use Exception;
+use yii\behaviors\AttributeBehavior;
+use yii\db\ActiveRecord;
+use yii\helpers\ArrayHelper;
+
+class User extends UserBase
+{
+    const SCENARIO_SAVE            = 'save';
+    const SCENARIO_UPDATE_PASSWORD = 'update_password';
+    const SCENARIO_AUTHENTICATE    = 'authenticate';
+
+    public $password;
+
+    public function scenarios(): array
+    {
+        $scenarios = parent::scenarios();
+
+        $scenarios[self::SCENARIO_DEFAULT] = [
+            'employee_id',
+            'first_name',
+            'last_name',
+            'display_name',
+            'username',
+            'email',
+            'active',
+            'locked',
+        ];
+
+        $scenarios[self::SCENARIO_UPDATE_PASSWORD] = ['password'];
+
+        $scenarios[self::SCENARIO_AUTHENTICATE] = ['username', 'password'];
+
+        return $scenarios;
+    }
+
+    public function rules(): array
+    {
+        return ArrayHelper::merge([
+            [
+                'active', 'default', 'value' => 'yes',
+            ],
+            [
+                'locked', 'default', 'value' => 'no',
+            ],
+            [
+                ['active', 'locked'], 'in', 'range' => ['yes', 'no'],
+            ],
+            [
+                'email', 'email',
+            ],
+            [
+                'password', 'required',
+                'on' => [self::SCENARIO_UPDATE_PASSWORD, self::SCENARIO_AUTHENTICATE],
+            ],
+            [
+                'password', 'string',
+            ],
+            [
+                'password',
+                $this->validatePassword(),
+                'on' => self::SCENARIO_AUTHENTICATE,
```

Comment:
I'm certainly not married to this implementation and appreciate the distinction of this sort of validation since there's no actual update to the database occurring.  It can be very tempting to utilize aspects of a framework to accomplish some task in a way that is out of scope or never intended by the authors...you know, the old "once I have a hammer everything starts looking like a nail."

I'm attempting to stick to a pattern I see emerging from the Yii docs that advocates this idea of strict adherence to the MVC architecture (which I realize can even have its own varied interpretations).  For everything so far, I rely on the model to establish all the "rules" whether or not they're related to data integrity like types or business rules like emails need to be a "valid".  The semantics of my thinking here is, "the password needs to match the hash" is similar to "the email needs to match an acceptable structure".  In my mind I made the semantical argument that a rule is rule and the model knows best how to ensure its integrity.

In the end, I believe this will boil down to stylistic choice and am comfortable refactoring it in a way that is more comfortable for you and @fillup if he feels the same way.  I'm assuming a refactoring would simply result in an explicit method call rather than relying on Yii's existing lifecycle, does that sound about right?